pkg/chunk/storage/caching_index_client.go

```diff
@@ -135,18 +150,29 @@ func (s *cachingIndexClient) QueryPages(ctx context.Context, queries []chunk.Ind
 		defer resultsMtx.Unlock()
 		keys := make([]string, 0, len(results))
 		batches := make([]ReadBatch, 0, len(results))
+		var cardinalityErr error
 		for key, batch := range results {
+			cardinality := int32(len(batch.Entries))
+			if cardinalityLimit > 0 && cardinality > cardinalityLimit {
+				batch.Cardinality = cardinality
+				batch.Entries = nil
```

Comment:
This is dangerous, this assumes that the cardinality limit is static, but it can be dynamically changed with a file. And given we cache the entries forever, this won't go away until we potentially rotate the caches.

Reply:
Fair point!  I don't want to cache very large entries (it will slow down the fetch, just to discard them).  Perhaps we should throw away empty cache entries?  Please file an issue.